### PR TITLE
[ASTPrinter] Restore printing generic requirement contexts

### PR DESCRIPTION
**Pull request by bitjammer**
cfe9e6a3defb0368c2ef9742b59bf1d83ca66ecf removed calls to pre/post
printing of PrintStructureKind::GenericRequirement, so SourceKit DocInfo
requests started droping the markers for generic requirements, causing
some weirdness with documentation rendering and post-processing.

Restore the calls to printStructPre/Post when printing generic
requirements.

rdar://problem/30561880

### CODE REVIEW

**Comment by bitjammer** [comment]
@swift-ci Please smoke test

File: include/swift/AST/ASTPrinter.h [hunk 1]
```diff
@@ -263,6 +263,14 @@ class ASTPrinter {
     printStructurePre(Kind, D);
   }
 
+  /// Make a callback to printStructurePost(), performing any necessary
+  /// bookkeeping.
+  void callPrintStructurePost(PrintStructureKind Kind,
+                              const Decl *D = nullptr) {
+    forceNewlines();
```
Comment (benlangmuir) [inline include/swift/AST/ASTPrinter.h]: How do we get in a situation where we want to print newlines before the "post" callback?  Usually we force newlines before the "pre" so that the callback happens at the point where the contents is being inserted.  For "post" this seems backwards to me.

File: test/Syntax/RoundTrip/fallthrough.swift [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// RUN: %round-trip-syntax-test -f %s --swift-syntax-test %swift-syntax-test
+fallthrough
```
Comment (benlangmuir) [inline test/Syntax/RoundTrip/fallthrough.swift]: Unrelated change?

Comment (bitjammer) [inline test/Syntax/RoundTrip/fallthrough.swift]: Yeah - I nuked it in the 3.1 PR.

File: include/swift/AST/ASTPrinter.h (hunk 1 above)
Comment (bitjammer) [inline include/swift/AST/ASTPrinter.h]: Ah, ok. I was wondering why we didn't have this - I didn't realize we were sticking to a pattern.

**Comment by bitjammer** [comment]
Making those changes on master - will cherry-pick here shortly. Stand by.
